File: apps/desktop/src-tauri/src/commands/queue.rs
use tauri::AppHandle;

use crate::{
    application::queue::{available_finish_actions, execute_system_action},
    domain::QueueFinishAction,
    error::AppError,
};

#[tauri::command]
pub fn available_queue_finish_actions() -> Vec<QueueFinishAction> {
    available_finish_actions()
}

#[tauri::command]
pub fn perform_queue_finish_action(
    app: AppHandle,
    action: QueueFinishAction,
) -> Result<(), AppError> {
    match action {
        QueueFinishAction::Exit => {
            app.exit(0);
            Ok(())
        }
        QueueFinishAction::SystemSleep | QueueFinishAction::SystemShutdown => {
            execute_system_action(action)
        }
        QueueFinishAction::Nothing => Ok(()),
    }
}

File: apps/desktop/src/lib/tauri/queue.ts
import { invoke } from "@tauri-apps/api/core";

import type { QueueFinishAction } from "./queue.types";

const QUEUE_FINISH_ACTIONS: readonly QueueFinishAction[] = [
  "exit",
  "systemSleep",
  "systemShutdown",
  "nothing",
];

function isQueueFinishAction(value: unknown): value is QueueFinishAction {
  return typeof value === "string" && QUEUE_FINISH_ACTIONS.includes(value as QueueFinishAction);
}

function normalizeQueueError(error: unknown): Error {
  if (error instanceof Error) return error;
  if (typeof error === "string") return new Error(error);
  if (typeof error === "object" && error !== null && "message" in error) {
    const message = (error as { message: unknown }).message;
    if (typeof message === "string") return new Error(message);
  }
  return new Error("The queue finish action could not be completed.");
}

export async function availableQueueFinishActions(): Promise<QueueFinishAction[]> {
  try {
    const value: unknown = await invoke("available_queue_finish_actions");
    if (!Array.isArray(value)) throw new Error("Invalid queue finish actions response.");
    return value.filter(isQueueFinishAction);
  } catch (error: unknown) {
    throw normalizeQueueError(error);
  }
}

export async function performQueueFinishAction(action: QueueFinishAction): Promise<void> {
  try {
    await invoke("perform_queue_finish_action", { action });
  } catch (error: unknown) {
    throw normalizeQueueError(error);
  }
}
